### PR TITLE
Pod task revamped

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Handle PodSpec definitions as the custom for pod tasks with task_type_version = 1

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/816

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

File: go/tasks/plugins/k8s/sidecar/sidecar.go [hunk 1]
```diff
@@ -73,45 +73,42 @@ type sidecarJob struct {
...
-		return nil, errors.Errorf(errors.BadTaskSpecification,
-			"invalid TaskSpecification [%v], Err: [%v]", task.GetCustom(), err.Error())
+	if task.TaskTypeVersion == 0 {
+		sidecarJob := sidecarJob{}
+		err := utils.UnmarshalStructToObj(task.GetCustom(), &sidecarJob)
+		if err != nil {
+			return nil, errors.Errorf(errors.BadTaskSpecification,
+				"invalid TaskSpecification [%v], Err: [%v]", task.GetCustom(), err.Error())
+		}
+		podSpec = *sidecarJob.PodSpec
```
Comment (EngHabu) [inline go/tasks/plugins/k8s/sidecar/sidecar.go:92]: Check if it's not nil first?

File: go/tasks/pluginmachinery/utils/marshal_utils.go [hunk 2]
```diff
@@ -68,3 +68,21 @@ func MarshalObjToStruct(input interface{}) (*structpb.Struct, error) {
 	}
 	return structObj, nil
 }
+
+// Don't use this if the unmarshalled obj is a proto message.
+func UnmarshalStructToObj(structObj *structpb.Struct, obj interface{}) error {
```
Comment (EngHabu) [inline go/tasks/pluginmachinery/utils/marshal_utils.go:73]: These are becoming confusing :D I think this naming is better than the other ones in the file... there are also a bunch of them here: https://github.com/flyteorg/flytestdlib/blob/519a271602eb6c88d58be3cf1576d03113beee31/utils/marshal_utils.go#L17

mind filing a housekeeping bug to clean that mess?

File: go/tasks/plugins/k8s/sidecar/sidecar.go [hunk 3]
```diff
@@ -73,45 +73,42 @@ type sidecarJob struct {
...
+		primaryContainerName = sidecarJob.PrimaryContainerName
+	} else {
+		err := utils.UnmarshalStructToObj(task.GetCustom(), &podSpec)
+		if err != nil {
+			return nil, errors.Errorf(errors.BadTaskSpecification,
+				"Unable to unmarshal task custom [%v], Err: [%v]", task.GetCustom(), err.Error())
+		}
+		primaryContainerName = task.GetConfig()[primaryContainerKey]
```
Comment (kumare3) [inline go/tasks/plugins/k8s/sidecar/sidecar.go:100]: what if this is missing?, check for that
```suggestion
		primaryContainerName, ok = task.GetConfig()[primaryContainerKey]
```

File: go/tasks/pluginmachinery/utils/marshal_utils.go (hunk 2 above)
Comment (katrogan) [reply go/tasks/pluginmachinery/utils/marshal_utils.go:73]: sure, done https://github.com/flyteorg/flyte/issues/828

File: go/tasks/plugins/k8s/sidecar/sidecar.go (hunk 3 above)
Comment (katrogan) [reply go/tasks/plugins/k8s/sidecar/sidecar.go:112]: done

File: go/tasks/plugins/k8s/sidecar/sidecar.go (hunk 1 above)
Comment (katrogan) [reply go/tasks/plugins/k8s/sidecar/sidecar.go:96]: done

**Comment by katrogan** [comment]
PTAL @EngHabu @kumare3

File: go/tasks/plugins/k8s/sidecar/sidecar.go [hunk 4]
```diff
@@ -73,45 +73,54 @@ type sidecarJob struct {
...
+		}
+		if len(task.GetConfig()) == 0 {
+			return nil, errors.Errorf(errors.BadTaskSpecification,
+				"invalid TaskSpecification, config needs to be non-empty and include missing [%s] key", primaryContainerKey)
+		}
+		if _, ok := task.GetConfig()[primaryContainerKey]; !ok {
+			return nil, errors.Errorf(errors.BadTaskSpecification,
+				"invalid TaskSpecification, config missing [%s] key in [%v]", primaryContainerKey, task.GetConfig())
+		}
+		primaryContainerName = task.GetConfig()[primaryContainerKey]
```
Comment (EngHabu) [inline go/tasks/plugins/k8s/sidecar/sidecar.go:112]:
```suggestion
		primaryContainerName, ok := task.GetConfig()[primaryContainerKey]
		if !ok {
			return nil, errors.Errorf(errors.BadTaskSpecification,
				"invalid TaskSpecification, config missing [%s] key in [%v]", primaryContainerKey, task.GetConfig())
		}
```

File: go/tasks/plugins/k8s/sidecar/sidecar.go [hunk 5]
```diff
@@ -73,45 +73,54 @@ type sidecarJob struct {
...
-	if err = json.Unmarshal(jsonObj, sidecarJob); err != nil {
-		return err
-	}
-
-	return nil
-}
-
 func (sidecarResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsCore.TaskExecutionContext) (client.Object, error) {
-	sidecarJob := sidecarJob{}
+	var podSpec k8sv1.PodSpec
```
Comment (EngHabu) [inline go/tasks/plugins/k8s/sidecar/sidecar.go:77]: nit: maybe make this a pointer, and just assign to it in the if and else below, then have the check `if podSpec == nil` outside the if taskTypeVersion condition?

File: go.sum [hunk 6]
```diff
@@ -228,6 +228,10 @@ github.com/fatih/color v1.10.0/go.mod h1:ELkj/draVOlAH/xkhN6mQ50Qd0MPOk5AAr3maGE
 github.com/fatih/structtag v1.2.0/go.mod h1:mBJUNpUnHmRKrKlQQlmCrh5PuhftFbNv8Ys4/aAZl94=
 github.com/flyteorg/flyteidl v0.18.15 h1:sXrlwTRaRjQsXYMNrY/S930SKdKtu4XnpNFEu8I4tn4=
 github.com/flyteorg/flyteidl v0.18.15/go.mod h1:b5Fq4Z8a5b0mF6pEwTd48ufvikUGVkWSjZiMT0ZtqKI=
+github.com/flyteorg/flyteidl v0.18.17-0.20210311212639-f1fce8de38db h1:GEnhYOicHA8RZPcbQwd4wYrBah1npIEQgtgacHgegJY=
```
Comment (EngHabu) [inline go.sum:231]: `go mod tidy` ?

**Review by EngHabu — COMMENTED** [review]
a couple of nits... sorry

File: go/tasks/plugins/k8s/sidecar/sidecar.go (hunk 5 above)
Comment (katrogan) [reply go/tasks/plugins/k8s/sidecar/sidecar.go:77]: I can't make it a pointer! unmarshaling the struct into a nil pointer fails unfortunately https://golang.org/pkg/encoding/json/#InvalidUnmarshalError

File: go.sum (hunk 6 above)
Comment (katrogan) [reply go.sum]: done @EngHabu

File: go/tasks/plugins/k8s/sidecar/sidecar.go (hunk 5 above)
Comment (EngHabu) [reply go/tasks/plugins/k8s/sidecar/sidecar.go:77]: ah right 👍

**Review by EngHabu — APPROVED** [review]
thank you

**Pull request merged**